DOCS 522: improve sensor api and add model configuration pages (#981)
viamrobotics/docs: docs/components/sensor/_index.md

> ---

Reviewer: That's ok

title: "Sensor Component"
linkTitle: "Sensor"
weight: 70
draft: false
type: "docs"
description: "A device that sends information about the outside world to the computer controlling a robot."
tags: ["sensor", "components"]

> icon: "img/components/sensor.png"

Reviewer: ```suggestion
no_list: true
icon: "img/components/sensor.png"
```

Reviewer: We don't have the list of cards at the end of the other components

Author: sounds good, done!

# SME: #team-bucket
---

A *sensor* is a device that can measure information about the outside world.

> Adding a sensor component to your robot allows the information the sensor measures to be sent to the computer controlling the robot.

Author: yes, done!

Most robots with a sensor need at least the following hardware:

- A [board](/components/board/)
- Depending on your sensor's output type (analog or digital), an analog-to-digital converter (ADC) may be necessary to allow the sensor to communicate with the board

## Configuration

Supported sensor models include:

| Model | Description |
| ----- | ----------- |
| [`fake`](fake) | A model used for testing, with no physical hardware. |
| [`ultrasonic`](ultrasonic) | [HC-S204 ultrasonic distance sensor](https://www.sparkfun.com/products/15569) |
| [`bme280`](bme280) | [BME280 environmental sensor](https://www.adafruit.com/product/2652) |
| [`ds18b20`](ds18b20) | [DallasTemperature DS18B20 digital temperature sensor](https://www.adafruit.com/product/381) |
| [`power_ina219`](power_ina219) | [INA219 current sensor](https://www.amazon.com/dp/B07QJW6L4C) |
| [`renogy`](renogy)| [Renogy battery temperature sensor](https://www.amazon.com/Renogy-Battery-Temperature-Sensor-Controllers/dp/B07WMMJFWY) |
| [`sensirion-sht3xd`](sensirion-sht3xd) | [Sensirion SHT3x-DIS temperature and humidity sensor](https://www.adafruit.com/product/2857) |

> Want to use another model of sensor to build your robot?

Author: done! oops haha

You can implement a model of sensor that is not natively supported by Viam by creating and registering your own model of a sensor.

> This allows you to have the same access and control of the sensor through Viam as you would if it was already a registered model.

Author: done!

> See [the Viam Labs GitHub](https://github.com/viam-labs/wifi-sensor) for an example of using the Viam Go SDK to implement a [WiFi strength sensor](https://github.com/viam-labs/wifi-sensor/blob/main/linuxwifi/linuxwifi.go), and [the Viam Robotics GitHub](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples/module) for an example of creating a custom component module with the Viam Python SDK.

Reviewer: How about

```suggestion
For an example of creating a custom component, see a [WiFi strength sensor built with the Viam Go SDK](https://github.com/viam-labs/wifi-sensor/blob/main/linuxwifi/linuxwifi.go) or the same WiFI strength sensor [implemented with the Viam Python SDK](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples/module).
```

Author: making some changes here but unfortunately it's not quite the same example!

> {{% alert title="Note" color="note" %}}

Reviewer: Sorry I missed this before - but I think the note might be better be above "most robots with a sensor..."?

Author: 👀 good point, done!

> See [here](https://github.com/viam-labs/wifi-sensor) for an example of using the Viam Go SDK to implement a [WiFi stength sensor](https://github.com/viam-labs/wifi-sensor/blob/main/linuxwifi/linuxwifi.go), and [here](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples/module) for an example of creating a custom component module with the Viam Python SDK.

Author: done I think!

> Viam has two component types defined separately from *sensor* that you can use to implement sensors with specific functions.

Author: Done!

> 1. [Movement sensor](/components/movement-sensor/) is defined for Global Positioning System (GPS) units, inertial measurement units (IMUs), and other sensors that detect position, velocity, and acceleration.

Reviewer: ```suggestion
1. [Movement sensors](/components/movement-sensor/) for Global Positioning System (GPS) units, inertial measurement units (IMUs), and other sensors that detect position, velocity, and acceleration.
```

> 2. [Encoder](/components/encoder/) is defined for sensors that can detect speed and direction of rotation of a motor or a joint.

Reviewer: ```suggestion
2. [Encoders](/components/encoder/) for sensors that can detect speed and direction of rotation of a motor or a joint.
```

> Viam also has an [encoder component](/components/encoder/) that is distinct from sensor.

Author: yes, done!

{{% /alert %}}

## Control your sensor with Viam's client SDK libraries

To get started using Viam's SDKs to connect to and control your robot, go to your robot's page on [the Viam app](https://app.viam.com), navigate to the **CODE SAMPLE** tab, select your preferred programming language, and copy the sample code generated.

When executed, this sample code will create a connection to your robot as a client.

> Then, adding API method calls into the code, as shown in the examples below, will allow you to control your robot programmatically.

Reviewer: ```suggestion
Then, you to control your robot programmatically by adding API method calls as shown in the following examples. 
```

^ a bit more direct

Author: Done!

These examples assumes you have a sensor called "my_sensor" configured as a component of your robot.
If your sensor has a different name, change the `name` in the code.

## API

The sensor component supports the following methods:

| Method Name | Description |
| ----------- | ----------- |

> [Readings](#readings) | Get the measurements or readings that this sensor provides. |

Reviewer: ```suggestion
| [Readings](#readings) | Get the measurements or readings that this sensor provides. |
```
